numpy/_core/src/umath/string_fastsearch.h

```diff
@@ -56,33 +55,156 @@
 #define MEMCHR_CUT_OFF 15
 
 
-template <ENCODING enc>
+template <typename char_type>
+struct CheckedIndexer {
+    char_type *buffer;
+    size_t length;
+
+    CheckedIndexer<char_type>()
+    {
+        buffer = NULL;
+        length = 0;
+    }
+
+    CheckedIndexer<char_type>(char_type *buf, size_t len)
+    {
+        buffer = buf;
+        length = len;
+    }
+
+    char_type
+    operator*()
+    {
+        return *(this->buffer);
+    }
+
+    char_type
+    operator[](size_t index)
+    {
+        if (index >= this->length) {
+            return (char_type) 0;
+        }
+        return this->buffer[index];
+    }
+
+    CheckedIndexer<char_type>
+    operator+(size_t rhs)
+    {
+        if (rhs > this->length) {
+            rhs = this->length;
+        }
+        return CheckedIndexer<char_type>(this->buffer + rhs, this->length - rhs);
+    }
+
+    CheckedIndexer<char_type>&
+    operator+=(size_t rhs)
+    {
+        if (rhs > this->length) {
+            rhs = this->length;
+        }
+        this->buffer += rhs;
+        this->length -= rhs;
+        return *this;
+    }
+
+    CheckedIndexer<char_type>
+    operator++(int)
+    {
+        *this += 1;
+        return *this;
+    }
+
+    CheckedIndexer<char_type>&
+    operator-=(size_t rhs)
+    {
+        this->buffer -= rhs;
+        this->length += rhs;
+        return *this;
+    }
+
+    CheckedIndexer<char_type>
+    operator--(int)
+    {
+        *this -= 1;
+        return *this;
+    }
+
+    std::ptrdiff_t
+    operator-(CheckedIndexer<char_type> rhs)
+    {
+        return this->buffer - rhs.buffer;
+    }
+
+    CheckedIndexer<char_type>
+    operator-(size_t rhs)
+    {
+        return CheckedIndexer(this->buffer - rhs, this->length + rhs);
+    }
+
+    int
+    operator>(CheckedIndexer<char_type> rhs)
+    {
+        return this->buffer > rhs.buffer;
+    }
+
+    int
+    operator>=(CheckedIndexer<char_type> rhs)
+    {
+        return this->buffer >= rhs.buffer;
+    }
+
+    int
+    operator<(CheckedIndexer<char_type> rhs)
+    {
+        return this->buffer < rhs.buffer;
+    }
+
+    int
+    operator<=(CheckedIndexer<char_type> rhs)
+    {
+        return this->buffer <= rhs.buffer;
+    }
+
+    int
+    operator==(CheckedIndexer<char_type> rhs)
+    {
+        return this->buffer == rhs.buffer;
+    }
+};
+
+
+template <typename char_type>
 inline Py_ssize_t
-findchar(Buffer<enc> s, Py_ssize_t n, npy_ucs4 ch)
+findchar(CheckedIndexer<char_type> s, Py_ssize_t n, char_type ch)
```

Comment:
Would it make sense to put all the functions here inside `CheckedIndexer` as static methods to avoid having to call the constructor and the function in two different lines? Maybe also rename it to something like `CheckIndexFinder`, if we do so.

Reply:
I don't see how that makes things simpler, since you'd still need to call the member function after creating the instance. Maybe I'm misunderstanding, can you illustrate the API you're thinking of a little more?

Reply:
I _think_ that you can just create a `CheckedIndexer` for the buffer to search in, and not the buffer to search for. So in every ufunc that will need to call `fastsearch`/`findchar` etc., you could create a `CheckedIndexer` in the beginning of the ufunc, and then do `ci.findchar(ch, len)`. In my mind this seems a bit clearer, but, obviously, feel free to disregard.

Reply:
That doesn't work for UTF8, the string lengths aren't uniform so we need a new indexer for each string. I could thread the `Indexer` through from the ufunc loops so the fixed-width strings could re-use it, but i'd still need to mutate the `Indexer` for every string to change the underlying character buffer. To me, this feels simpler, at the cost of a little bit of extra overhead for fixed-width strings.